src/dbnode/storage/bootstrap/bootstrapper/base.go

```diff
@@ -76,133 +74,252 @@ func (b baseBootstrapper) String() string {
+			if err != nil {
+				return bootstrap.NamespaceResults{}, err
+			}
+
+			currNamespace.IndexRunOptions.ShardTimeRanges = indexAvailable
+		}
+
+		// Set the namespace options for the current bootstrapper source.
+		curr.Namespaces.Set(id, currNamespace)
+
+		// Log the metadata about bootstrapping this namespace based on
+		// the availability returned.
+		b.logShardTimeRanges("bootstrap from source ready after availability query",
+			logFields, currNamespace)
 	}
-	b.log.Info("bootstrapping from source starting", logFields...)
 
 	nowFn := b.opts.ClockOptions().NowFn()
 	begin := nowFn()
 
-	currStatus, currErr = step.runCurrStep(currRanges)
+	b.log.Info("bootstrap from source started", logFields...)
+	currResults, err := b.src.Read(curr)
 
 	logFields = append(logFields, zap.Duration("took", nowFn().Sub(begin)))
-	if currErr != nil {
-		logFields = append(logFields, zap.Error(currErr))
-		b.log.Info("bootstrapping from source completed with error", logFields...)
-	} else {
-		logFields = append(logFields, currStatus.logFields...)
-		b.log.Info("bootstrapping from source completed successfully", logFields...)
+	if err != nil {
+		errorLogFields := append(logFieldsCopy(logFields), zap.Error(err))
+		b.log.Error("error bootstrapping from source", errorLogFields...)
+		return bootstrap.NamespaceResults{}, err
 	}
 
-	wg.Wait()
-	if err := xerrors.FirstError(currErr, nextErr); err != nil {
-		return err
+	// Determine the unfulfilled and the unattempted ranges to execute next.
+	next, err := b.logSuccessAndDetermineCurrResultsUnfulfilledAndNextBootstrapRanges(namespaces,
+		curr, currResults, logFields)
+	if err != nil {
+		return bootstrap.NamespaceResults{}, err
 	}
 
-	fulfilledRanges := result.ShardTimeRanges{}
-	fulfilledRanges.AddRanges(currStatus.fulfilled)
-	fulfilledRanges.AddRanges(nextStatus.fulfilled)
-	unfulfilled := totalRanges.Copy()
-	unfulfilled.Subtract(fulfilledRanges)
+	// Unless next bootstrapper is required, this is the final results.
+	finalResults := currResults
 
-	step.mergeResults(unfulfilled)
+	// If there are some time ranges the current bootstrapper could not fulfill,
+	// that we can attempt then pass it along to the next bootstrapper.
+	if next.Namespaces.Len() > 0 {
+		nextResults, err := b.next.Bootstrap(next)
+		if err != nil {
+			return bootstrap.NamespaceResults{}, err
+		}
 
-	unattemptedNextRanges := currRanges.Copy()
-	unattemptedNextRanges.Subtract(currStatus.fulfilled)
-	if !nextAttempted {
-		// If we have never attempted the next time range then we want to also
-		// attempt the ranges we didn't even try to attempt
-		unattemptedNextRanges.AddRanges(nextRanges)
+		// Now merge the final results.
+		for _, elem := range nextResults.Results.Iter() {
+			id := elem.Key()
+			currNamespace := elem.Value()
+
+			finalResult, ok := finalResults.Results.Get(id)
+			if !ok {
+				return bootstrap.NamespaceResults{},
+					fmt.Errorf("expected result for namespace: %s", id.String())
```

Comment:
Is there any chance that this bootstrapper may just not have the given namespace?